HACKING.md

```diff
@@ -551,6 +551,222 @@ Common instructions for porting:
+    on the page.) The second setting will tell you in the profiler which
+    props changed to trigger those updates.
+
+    Once you indentify which props are triggering unnecessary updates,
+    it's generally just a matter of making them more specific as per above,
+    or caching them with the `useMemo` and `useCallback` hooks. Another
+    situation might be where you are passing a prop which is used in a
+    callback (say, an event handler), but doesn't actually affect the
+    rendered output at all. Consider passing that prop as a
+    [ref](https://react.dev/reference/react/useRef) instead,
+    since the ref's identity will remain stable.
+
+  * The second issue with top-down data flow is related to organization: if
+    there's only one reducer function, then how do we avoid it becoming a
+    mess, managing tons of actions from many different files? And how do we
+    make it clear what component an action is dispatched from?
+
+    Even though there's only one "real" reducer, we don't have to make it
+    handle every single action in the app on its own. We can still define
+    separate reducers for each child component, and simply call them from the
+    parent reducer.
+
+    A general outline of the types and functions you'd typically define in
+    a component file are described below. These are specific to the file's
+    component, and handle all the actions/state used by that component.
+
+      * `type ActionT`: A union type that lists all the actions that the
+        component can `dispatch`. Each action should be an object with a
+        `type` field that names the action, e.g. `add-item`, `show-help`,
+        etc., along with any additional data needed by the action.
+
+      * `type StateT`: A read-only type that describes the state which is
+        needed to display the component and dispatch actions.
+
+      * `type PropsT`: In many cases, the only props you need are `dispatch`
+        (`(ActionT) => void`) and `state` (`StateT`), though it also makes
+        sense to pass static values that never change as props instead of
+        including them in `state`. If you need these props in the `reducer`
+        function, having them in `state` is convenient, though.
+
+      * `function createInitialState(...)`: Builds a `StateT`. The arguments
+        to this function are up to you and depend on what properties you need
+        to initialize.
+
+      * `function reducer(state: StateT, action: ActionT)`: Sets up a switch
+        statement over `action.type` and returns an updated `StateT`.
+
+        `StateT` should be deeply read-only. To make modifications, first
+        make a copy:
+
+        ```js
+        const newState = {...state};
+        // ... and while handling some action:
+        newState.someList = [...newState.someList, newItem];
+        ```
+
+        If you need to make complex updates to deeply-nested properties,
+        use the `mutate-cow` library.
```

Comment:
There used to be a sentence here suggesting to use `createFastObjectCloneFunction`, but that should rarely be necessary unless it's performance-critical code. I replaced it with a suggestion to use mutate-cow.

Reply:
:cow: